### PR TITLE
update moment module from testcafe/node_modules/ (closes #1750)

### LINKED ISSUE #1750: Running sample test thrown an UnhandledPromiseRejectionWarning: this.moment.duration(...).format is not a function
### Are you requesting a feature or reporting a bug?
- Could be a bug.

### What is the current behavior?

After a fresh installation within my project directory I followed the main example of the Getting Started doc's paragraph.  
The test example runs correctly, but at the end of the test the browser hangs on a Testcafe page that said ***Connected*** with a loading spinner and an error is shown on the terminal:

```bash
UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: this.moment.duration(...).format is not a function
```

### What is the expected behavior?

Finished tests and close the browser without errors (?).

### How would you reproduce the current behavior (if this is a bug)?

Installed within my Ubuntu:
```bash
VERSION="16.04.2 LTS (Xenial Xerus)"
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

Inside a project with the command:
```bash
yarn add --dev testcafe
```
Running with a yarn script `yarn test:testcafe` that runs testcafe:
```bash
testcafe chrome test/testcafe
```

Within the `testcafe`  folder I have a `main.js` file for my test.

##### main.js
```js
import { Selector } from 'testcafe';

fixture `Getting Started`
    .page `https://devexpress.github.io/testcafe/example`;

test('My first test', async t => {
    await t
        .typeText('#developer-name', 'John Smith')
        .click('#submit-button');
});
```

### Specify your

* operating system: Ubuntu (as stated)
* testcafe version: 0.17.1
* node.js version: 8.1.0

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

**Comment by AlexKamaev** [comment]
@testcafe-build-bot \retest

File: src/reporter/moment-duration-format-loader.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+try {
+    sideMomentModulePath = resolveFrom(momentDurationFormatPath, 'moment');
+}
+catch (err) {
+    sideMomentModulePath = testcafeMomentModulePath;
+}
+
+if (testcafeMomentModulePath !== sideMomentModulePath) {
+    require(sideMomentModulePath);
```
Comment (AndreyBelym) [inline src/reporter/moment-duration-format-loader.js]: I think, you shouldn't require the side module at all.

File: src/reporter/moment-duration-format-loader.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    const testcafeMomentModuleCached = require.cache[testcafeMomentModulePath];
+    const siteMomentModuleCached     = require.cache[sideMomentModulePath];
+
+    require.cache[sideMomentModulePath] = testcafeMomentModuleCached;
+
+    require(momentDurationFormatPath);
+
+    if (testcafeMomentModuleCached)
```
Comment (AndreyBelym) [inline src/reporter/moment-duration-format-loader.js]: You must check for `sideMomentModuleCached` here.

File: src/reporter/plugin-host.js [hunk 3]
```diff
@@ -1,8 +1,7 @@
 import chalk from 'chalk';
 import indentString from 'indent-string';
 import { identity, escape as escapeHtml, assignIn } from 'lodash';
-import moment from 'moment';
-import 'moment-duration-format';
+import moment from './moment-duration-format-loader';
```
Comment (AndreyBelym) [inline src/reporter/plugin-host.js]: Let's name the loader module `moment-loader.js`, and move it to `src/utils`.

File: src/reporter/moment-duration-format-loader.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const resolveFrom              = require('resolve-from');
+const testcafeMomentModulePath = require.resolve('moment');
```
Comment (AndreyBelym) [inline src/reporter/moment-duration-format-loader.js]: I think, we also must be carefull with TestCafe-related module. Let's handle it with the following algorithm:
```
const originalModule = cache[modulePath];

delete cache[modulePath];

const module = require(modulePath);
// ...
// do stuff
// ...

if (originalModule)
    cache[modulePath] = originalModule;
else
    delete cache[originalModule];
```

File: src/reporter/moment-duration-format-loader.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+}
+catch (err) {
+    sideMomentModulePath = testcafeMomentModulePath;
+}
+
+if (testcafeMomentModulePath !== sideMomentModulePath) {
+    require(sideMomentModulePath);
+
+    const testcafeMomentModuleCached = require.cache[testcafeMomentModulePath];
+    const siteMomentModuleCached     = require.cache[sideMomentModulePath];
```
Comment (AndreyBelym) [inline src/reporter/moment-duration-format-loader.js]: const sideMomentModuleCached

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

**Comment by AndreyBelym** [comment]
\сс @helen-dikareva

File: src/utils/moment-loader.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    try {
+        return resolveFrom(sidePath, 'moment');
+    }
+    catch (err) {
+        return '';
+    }
+}
+
+function getModulesPaths () {
+    const durationFormatModulePath = require.resolve('moment-duration-format');
```
Comment (helen-dikareva) [inline src/utils/moment-loader.js]: Maybe we should move modules names into consts?

File: src/utils/moment-loader.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        ...getCachedAndCleanModules(mainMomentModulePath)
+    };
+}
+
+function getMomentModuleWithDurationFormatPatch () {
+    const modulesPaths  = getModulesPaths();
+    const momentModules = getMomentModules(modulesPaths);
+
+    if (modulesPaths.sideMomentModulePath && modulesPaths.sideMomentModulePath !== modulesPaths.mainMomentModulePath) {
```
Comment (helen-dikareva) [inline src/utils/moment-loader.js]: we can assign `const { sideMomentModulePath, mainMomentModulePath, durationFormatModulePath } = modulesPaths;` to make code more readable

File: src/utils/moment-loader.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        require.cache[modulesPaths.sideMomentModulePath] = momentModules.mainModule;
+
+        require(modulesPaths.durationFormatModulePath);
+
+        restoreInitialCacheState(momentModules.sideModule, modulesPaths.sideMomentModulePath);
+    }
+    else {
```
Comment (helen-dikareva) [inline src/utils/moment-loader.js]: we can use `else if ... ` here

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

**Pull request merged**